misc/python/materialize/elf.py

```diff
@@ -0,0 +1,35 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+from typing import BinaryIO
+
+# `stubgen -p elftools` doesn't work out of the box,
+# and manually writing stub files for all of `elftools`
+# seems like a large and error-prone project.
```

Comment:
For other stuff we just vendor a stubfile with just the tiny subset of the API that we need.

Reply:
That seems strictly worse than what I have now. For example, if the type of `n_desc` ever changes, at least our program will still be type-safe, because I'm dynamically casting it to `str`. With our own made-up stubs, nothing enforces that they will actually be kept in sync with upstream, so any additional type safety is just an illusion. Am I missing something?

Reply:
I spoke about this with Nikhil offline, we concluded that my point here was valid, but if we start using `pyelftools` in other places outside of this `elf.py` type-safe wrapper, it would be good to have a pyi. Here's a sketch of one:

```python
# Copyright Materialize, Inc. and contributors. All rights reserved.
#
# Use of this software is governed by the Business Source License
# included in the LICENSE file at the root of this repository.
#
# As of the Change Date specified in that file, in accordance with
# the Business Source License, use of this software will be governed
# by the Apache License, Version 2.0.

from typing import Any, BinaryIO, Iterable

class Note:
    @property
    def n_type(self) -> str: ...
    @property
    def n_name(self) -> str: ...
    @property
    def n_desc(self) -> str: ...

class NoteSection:
    def iter_notes(self) -> Iterable[Note]: ...

class ELFFile:
    def __init__(self, stream: BinaryIO): ...
    def get_section_by_name(self, name: str) -> Any: ...
```